### PR TITLE
feat: wildcard support for modules in usage file

### PR DESCRIPTION
**Pull request by hugorut**
This PR adds support for wildcard references in the Infracost usage file.

Users can now specify usage file references such as `module.mod[*].aws_lambda_function.test[*]`.

This functionality is enabled by the introduction of `schema.UsageMap` into the application. This builds a regexp from the usage file name (escaping all regular expression metacharacters apart from *) if a direct match is not made. This means we can support many layers of wildcard references, e.g: `module.mod[*].module.another[*].aws_lambda_function.test[*]`.

### CODE REVIEW

**Comment by hugorut** [comment]
@aliscott I'm not sure what we're thinking of doing with the `sync` functionality and if these wildcard changes need to change that as well. For now I've not touched it.

File: internal/schema/usage_data.go [hunk 1]
```diff
@@ -149,8 +151,64 @@ func (u *UsageData) CalcEstimationSummary() map[string]bool {
...
+	if err == nil {
+		val, ok := usage[parsedAddress.ResourceSpec.Type]
+		if ok {
+			data = val
+		}
+	}
+
+	if ud := usage[address]; ud != nil {
+		if data != nil {
+			MergeAttributes(data, ud)
```
Comment (hugorut) [inline internal/schema/usage_data.go:172]: It seems from the code that I extracted this from that we were only ever merging attributes from the `type` spec if this was a direct hit. I'm not sure if we should keep this behaviour, or merge the usage data if we also get a hit on the wildcard matches. I think we probably should, but I'm unsure why this wasn't done in the first place?

File: internal/schema/usage_data.go [hunk 2]
```diff
@@ -149,8 +151,64 @@ func (u *UsageData) CalcEstimationSummary() map[string]bool {
...
+		return ud
+	}
+
+	if strings.HasSuffix(address, "]") {
+		lastIndexOfOpenBracket := strings.LastIndex(address, "[")
+
+		if arrayUsageData := usage[fmt.Sprintf("%s[*]", address[:lastIndexOfOpenBracket])]; arrayUsageData != nil {
+			return arrayUsageData
+		}
+	}
```
Comment (hugorut) [inline internal/schema/usage_data.go:184]: technically we don't "need" this now with the regxp matching below, however this is more performant lookup. So I'd prefer to keep it, so we catch things before moving to the matching functionality. However, I'm ok to remove as well.

File: internal/schema/usage_data.go (hunk 1 above)
Comment (hugorut) [reply internal/schema/usage_data.go:172]: Discussed on zoom, this is a bug.

File: cmd/infracost/testdata/breakdown_terraform_sync_usage_file/breakdown_terraform_sync_usage_file.golden [hunk 3]
```diff
@@ -8,7 +8,7 @@ Project: infracost/infracost/cmd/infracost/testdata/breakdown_terraform_sync_usa
  └─ Insights queries data scanned                    Monthly cost depends on usage: $0.005 per GB   
                                                                                                     
  aws_cloudwatch_log_group.production_logs["media"]                                                  
- ├─ Data ingested                                    Monthly cost depends on usage: $0.50 per GB    
+ ├─ Data ingested                                             1,000  GB                     $500.00
```
Comment (hugorut) [inline cmd/infracost/testdata/breakdown_terraform_sync_usage_file/breakdown_terraform_sync_usage_file.golden:11]: This seems like this was a bug before that was fixed by this PR, the resource defaults are as follows:

```
resource_type_default_usage:
  aws_cloudwatch_log_group:
    monthly_data_ingested_gb: 1000.0
```

**Review by aliscott — APPROVED** [review]
Nice tests!

**Pull request merged**